python/paddle/distributed/passes/pass_base.py

```diff
@@ -0,0 +1,273 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+# 
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#     http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import six
+import sys
+from abc import ABC, abstractmethod
+from paddle.fluid.framework import program_guard, _apply_pass as _apply_cpp_pass
+
+
+class PassContext:
+    def __init__(self):
+        self._applied_passes = []
+        self._attrs = []
+
+    def set_attr(self, key, value):
+        self._attrs[key] = value
+
+    def get_attr(self, key, default=None):
+        return self._attrs.get(key, default)
+
+    @property
+    def passes(self):
+        return self._applied_passes
+
+    def _add_pass(self, pass_obj):
+        self._applied_passes.append(pass_obj)
+
+    def _pop_pass(self):
+        del self._applied_passes[-1]
+
+
+class PassBase(ABC):
+    _REGISTERED_PASSES = {}
+    _COMMON_RULES = []
+
+    @staticmethod
+    def _register(pass_name, pass_class):
+        assert issubclass(pass_class, PassBase)
+        PassBase._REGISTERED_PASSES[pass_name] = pass_class
+
+    def __init__(self):
+        self._attrs = {}
+
+    def set_attr(self, key, value):
+        self._attrs[key] = value
+        return self
+
+    def get_attr(self, key, default=None):
+        return self._attrs.get(key, default)
+
+    @abstractmethod
+    def _check_self(self):
+        pass
+
+    @abstractmethod
+    def _check_conflict(self, other_pass):
+        pass
+
+    def _check_conflict_including_common_rules(self, other_pass):
+        return self._check_conflict(other_pass) and all(
+            [r(other_pass, self) for r in PassBase._COMMON_RULES])
+
+    def apply(self, main_programs, startup_programs, context=None):
```

Comment:
what about to have 4 apply methods: apply_forward, apply_backward, apply_optimize, and apply(include above three) 
1: it would more nature to for the implementation of amp & recompute pass, since they will modify the forward and backward program with different logic.
2: if we want to generate a program with forward ops only (for inference ), it will more convenient we could call amp.apply_forward for this program instead of amp.apply